Print sentry

BacLuc opened the pull request:
## Changes

1. print: report caught error in pdfGeneratorNativeChrome.js

If you catch an error, you also have to report it.
Else it does not end up in sentry.

Issue: #3471

2. print: report console errors to sentry

That we also see errors when server side rendering the html for the pdf.

Issue: #3471

3. print: make timeout for html rendering configurable

A higher config value reduces the " ERROR  Navigation timeout of $CONFIG_VALUE ms exceeded"

Issue: #3472

4. print: make timeout for pdf rendering configurable

A higher config value reduces the " ERROR  Protocol error (Page.printToPDF): Printing failed  "

5. (optional): .helm/browserless: make connectionTimeout configurable

BROWSERLESS_CONNECTION_TIMEOUT_MS and PRINT_RENDER_PDF_TIMEOUT_MS were related
on the machine of @BacLuc

6. print: also log fetch errors which will then be sent to sentry

They are logged in development mode, but not in production mode by default.
Found the hooks by searching for "callHook("" in node_modules/\@nuxt,
and this hook was called in production mode after the fetch errors happened
with the fetch errors in the nuxt object.
This also helped: https://github.com/nuxt/nuxt/issues/7742#issuecomment-743205834

7. print: make read-timeout for ingress configurable

If we print a large pdf and this takes longer than 60s,
the ingress-nginx controller retries the request.
Default is 60s, https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout

## Manual Tests
- [x] Not setting the new ENV Variables for the deployment leads to the defaults defined in the js code being used. (And the deployment still works)
- [x] Errors in the middleware appear in sentry: https://ecamp.sentry.io/issues/4201963682/?environment=pr3481.ecamp3.ch&project=5603959&query=is%3Aunresolved&referrer=issue-stream&stream_index=0
- [x] Render errors in the SSR appear in sentry
- [x] Fetch errors in the SSR appear in sentry
- [x] Increasing PRINT_RENDER_HTML_TIMEOUT allows to load more (or longer taking) html
- [x] Increasing PRINT_RENDER_PDF_TIMEOUT (With BROWSERLESS_CONNECTION_TIMEOUT_MS) allows to render a pdf which takes longer
- [x] Changing the ENV Variables for the deployment propagates to the containers

## Other findings

BacLuc committed:
print: report caught error in pdfGeneratorNativeChrome.js

If you catch an error, you also have to report it.
Else it does not end up in sentry.

Issue: #3471

BacLuc committed:
print: report console errors to sentry

That we also see errors when server side rendering the html for the pdf.

Issue: #3471

BacLuc committed:
print: make timeout for html rendering configurable

A higher config value reduces the " ERROR  Navigation timeout of $CONFIG_VALUE ms exceeded"

Issue: #3472

BacLuc committed:
print: make timeout for pdf rendering configurable

A higher config value reduces the " ERROR  Protocol error (Page.printToPDF): Printing failed  "

BacLuc committed:
.helm/browserless: make connectionTimeout configurable

BROWSERLESS_CONNECTION_TIMEOUT_MS and PRINT_RENDER_PDF_TIMEOUT_MS were related
on the machine of @BacLuc

usu commented on print/server-middleware/routes/pdfGeneratorNativeChrome.js:
```diff
@@ -121,28 +123,32 @@ router.use('/pdfChrome', async (req, res) => {
-    if (browser) {
-      browser.disconnect()
-    }
+    try {
```
The nested try-finally block makes the code a bit difficult to read.

Wouldn't it be easier to just put the `process.sentry.captureException(error)` at the beginning of the catch-block, so we are sure that the error is immediately sent to sentry and we can then do the rest of the error handling? Thus avoid the nesting of try-catch-try?

BacLuc replied:
done

usu commented on .github/workflows/deployment-stage-prod.yml:
```diff
@@ -86,6 +86,8 @@ jobs:
             --set php.oauth.jubladb.baseUrl='${{ secrets.OAUTH_JUBLADB_BASE_URL }}' \
             --set frontend.sentryDsn='${{ secrets.FRONTEND_SENTRY_DSN }}' \
             --set print.sentryDsn='${{ secrets.PRINT_SENTRY_DSN }}' \
+            --set print.renderHTMLTimeoutMs='${{ env.PRINT_RENDER_HTML_TIMEOUT_MS }}' \
```
Shouldn't this be vars instead of env?

Do we actually intend to configure the timeouts differently for the various environments? If not, I'd also be ok to skip these 2 lines to keep the deployment command simpler.

carlobeltrame replied:
Not sure what you mean by "vars". If you meant default values in values.yaml, I would agree.

BacLuc replied:
yes, they are vars, my bad.

I would like to be able to verify this before deploying to stage.
(and i think it would be better to unify the 2 deployment workflows, that for the configurable values it is always clear what you have to do.
Then you don't have to think where i make what configurable, but that you make it configurable and always in the same way.
It is important that everything works if the vars or the secret is not available if this is possible.)
In my experience you don't know in advance what should be configurable.

usu replied:
> In my experience you don't know in advance what should be configurable.

Not sure I got this last point. All parameters in `values.yaml` are configurable and hence `values.yaml` serves as the documentation what can be configured via `helm` command (for example for manual deployments).

I would object that we need to replicate all parameters which are theoretically configurable in our deployment scripts (they would get really really large; values.yaml has more than 200 lines at the moment). So unless we actually intend to have different values between dev/stage/prod I wouldn't list them in the deployment scripts.

If the intention is to play around with different values I'm ok. We can have them in and maybe remove later if we find good default values.

usu reviewed (commented):
Nice, thanks for digging into this 👍

BacLuc committed:
print: also log fetch errors which will then be sent to sentry

They are logged in development mode, but not in production mode by default.
Found the hooks by searching for "callHook("" in node_modules/@nuxt,
and this hook was called in production mode after the fetch errors happened
with the fetch errors in the property.
This also helped: https://github.com/nuxt/nuxt/issues/7742#issuecomment-743205834

BacLuc committed:
print: make read-timeout for ingress configurable

If we print a large pdf and this takes longer than 60s,
the ingress-nginx controller retries the request.
Default is 60s, https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout

usu commented on print/server-middleware/routes/pdfGeneratorNativeChrome.js:
```diff
@@ -33,6 +33,14 @@ function measurePerformance(msg) {
 }
 
+if (!process.sentry) {
+  process.sentry = {
```
Hmm, not sure I like this way of overriding `process.sentry`. What if some other code somewhere checks for `process.sentry` at one point in the future, and then gets confused, because `process.sentry` it's not an actual sentry object.

If you don't want to have the if-clause below, wouldn't it be cleaner to have a `captureException` method here which is responsible to either raise a `sentry.captureException` or simply log to the console, depending on whether sentry is available or not?

usu reviewed (approved):
Minor comment below. Rest looks good to me

carlobeltrame reviewed (approved):
I was able to trigger a Sentry error by printing a PDF containing 30x the programme of GRGR: https://ecamp.sentry.io/issues/4271516151/?environment=pr3481.ecamp3.ch&project=5603959&query=is%3Aunresolved&referrer=issue-stream&stream_index=0
